graph_train.py

```diff
@@ -0,0 +1,178 @@
+import os
+import argparse
+import torch
+from collections import defaultdict
+import random
+
+import nltk
+import torch_geometric.data as geom_data
+# nltk.download('reuters')
+# from nltk.corpus import reuters
+from datasets.reuters_graph_datasets import R8, R52
+from torch_geometric.data import Data
+import torch.nn.functional as F
+from torch_geometric.nn import GCNConv
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks import ModelCheckpoint
+
+
+from datasets.graph_utils import PMI, tf_idf_mtx
+
+device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")
+
+class Net(torch.nn.Module):
+    def __init__(self, num_nodes):
+        super(Net, self).__init__()
+        self.conv1 = GCNConv(num_nodes, 200)
+        self.conv2 = GCNConv(200, 8)
+
+    def forward(self, data):
+        x, edge_index, edge_weight = data.x, data.edge_index, data.edge_attr
+        x = self.conv1(x, edge_index, edge_weight)
+        x = F.relu(x)
+        x = F.dropout(x, training=self.training)
+        x = self.conv2(x, edge_index, edge_weight)
+        return x
+
+class Graph_model(pl.LightningModule):
+    def __init__(self, num_nodes):
+        super().__init__()
+        device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")
+        self.model = Net(num_nodes).to(device)
+        self.save_hyperparameters()
+    
+    def forward(self, data, mode):
+        out = self.model(data)
+        if mode=='train':
+            mask = data.train_mask
+        elif mode=='val':
+            mask = data.val_mask
+        elif mode=='test':
+            mask = data.test_mask
+        loss = F.cross_entropy(out[mask], data.y[mask])
+        # loss = F.nll_loss(out[data.train_mask], data.y[data.train_mask])
+        class_predictions = torch.argmax(out, dim=1)
+        correct = (class_predictions[mask] == data.y[mask]).sum().item()
+        accuracy = correct / mask.sum()
+        # if math.isnan(loss.item()):
+        #     print()
+        return loss, accuracy
+    
+    def configure_optimizers(self):
+        self.optimizer = torch.optim.Adam(self.parameters(), lr=0.01)
+        return self.optimizer
+
+    def training_step(self, batch, batch_idx):
+        loss, acc = self.forward(batch, mode='train')
+        self.log("train_loss", loss, on_step=False, on_epoch=True)
+        self.log("train_acc", acc, on_step=False, on_epoch=True)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        loss, acc = self.forward(batch, mode='val')
+        self.log("val_loss", loss)
+        self.log("val_acc", acc)
+
+    def test_step(self, batch, batch_idx):
+        loss, acc = self.forward(batch, mode='test')
+        self.log("test_loss", loss)
+        self.log("test_acc", acc)
+
+class GenerateCallback(pl.Callback):
+    def __init__(self):
+        """
+        Args:
+            gamma (float, optional): [description]. Defaults to 0.2.
```

Comment:
gamma?

Reply:
This was copied from my first assignment. File should be deleted anyways